src/utils/map.hpp

```diff
@@ -0,0 +1,968 @@
+#pragma once
+#include <functional>
+#include <iostream>
+#include <iomanip>
+
+
+
+namespace nstd
+{
+
+    /* FORWARD DECLARATIONS*/
+
+    template <typename key>
+    class hash;
+
+    template <typename key, typename value, typename Hash>
+    class map;
+
+    template <class k, class v, class Hash>
+    std::ostream& operator<<(std::ostream& out, const map<k, v, Hash>& d);
+
+    unsigned int* SieveOfEratosthenes(int n);
+
+    /*CLASSES & STRUCTS */
+
+    /**
+     * @class nstd::map class
+     *
+     * A data structure class that is meant to hold key value pairs and
+     * hash with the key to retrieve the value. This map is also iteratable
+     * meaning that you can start at the beginning or end and read through it
+     * as if it was a list. In the context of our project this class will relate
+     * team id's with a stadium and stadium id's with a team.
+     *
+     * This class also supports custom hashing, but the functions need to be in this
+     * format static int staticHashAlgo(key k,int j,int capacity) where the returned
+     * int from the function needs to follow this restriction 0 < hash < capacity.
+     *
+     * @code{.cpp}
+     * nstd::map<int,std::string> myMap;
+     * //will create a new key value in map if not present
+     * //othewise reassigns value at key 2
+     * myMap[2] = "Hi";
+     * //same if not present will create new value with default constructor
+     * // at key 8 otherwise returns value
+     * std::cout << myMap[8];
+     * @endcode
+     *
+     * In the code above, a new nstd::map is instantiated with the key being an int
+     * and the value being a std::string. We then assign Hi into a newly created value
+     * location. And we readout whatever is in key location 8, in our case nothing so
+     * it returns a default constructed value.
+     */
+    template <typename key, typename value, typename Hash = hash<key>>
+    class map
+    {
+    private:
+        struct node;
+    public:
+        class iterator;
+
+        /* Constructors, Destructor, and Assignment */
+        map();
+        map(key[],value[],const int);
+        map(const map&);
+        map(map&&);
+        map& operator=(const map&);
+        ~map();
+
+        /* Accessing */
+        iterator find(key);
+        value& operator[] (const key&);
+        value& operator[] (key&&);
+        friend std::ostream& operator<< <>(std::ostream&, const map<key,value,Hash>&);
+
+        /* Mutation Operations */
+        iterator insert(key, value);
+        std::pair<key,value> remove(key);
+        void reserve(unsigned int);
+
+        /* Iterator Getters */
+        iterator begin() noexcept;
+        iterator end() noexcept;
+
+        /* Map information*/
+        int size() const {return m_numOfElems;}
+        int capacity() const {return m_capacity;}
+        bool isEmpty() const {return (m_numOfElems == 0);}
```

Comment:
`empty()` instead of `isEmpty()` or else bjarne might get mad at us 😡 
You also have redundant parenthesis `return m_numOfElems == 0;`